XenAdmin/SettingsPanels/USBEditPage.cs

```diff
@@ -0,0 +1,234 @@
+using XenAdmin.Core;
+
+namespace XenAdmin.SettingsPanels
+{
+    public partial class USBEditPage : XenTabPage, IEditPage
+    {
+        private VM _vm;
+
+        public USBEditPage()
+        {
+            InitializeComponent();
+        }
+
+        #region override IEditPage
+
+        public bool ValidToSave
+        {
+            get { return true; }
+        }
+
+        public bool HasChanged
+        {
+            get { return true; }
+        }
+
+        public override string Text
+        {
+            get { return Messages.USB_EDIT_TEXT; }
+        }
+
+        public string SubText
+        { 
+            get
+            {
+                if (_vm != null)
+                {
+                    if (_vm.VUSBs.Count == 0)
+                        return Messages.USB_EDIT_SUBTEXT_NODEVICES;
+                    else if (_vm.VUSBs.Count == 1)
+                        return Messages.USB_EDIT_SUBTEXT_1_DEVICE;
+                    else
+                        return string.Format(Messages.USB_EDIT_SUBTEXT_MULTIPLE_DEVICES, _vm.VUSBs.Count);
+                }
+                else
+                    return "";
+            }
+        }
+
+        public Image Image
+        {
+            get { return Resources._001_Tools_h32bit_16; }
+        }
+
+        public AsyncAction SaveSettings()
+        {
+            return null;
+        }
+
+        public void SetXenObjects(IXenObject orig, IXenObject clone)
+        {
+            Trace.Assert(clone is VM);
+
+            _vm = (VM)clone;
+            if (Connection == null)
+                Connection = _vm.Connection;
+
+            BuildList();
+        }
+
+        public bool InBuildList = false;
+        private void BuildList()
+        {
+            Program.AssertOnEventThread();
+
+            if (InBuildList)
+                return;
+
+            InBuildList = true;
+
+            try
+            {
+                List<VUSB> vusbs = _vm.Connection.ResolveAll(_vm.VUSBs);
+
+                var VmUsbRowToAdd = new List<VMUsbRow>();
+                foreach (VUSB vusb in vusbs)
+                {
+                    VmUsbRowToAdd.Add(new VMUsbRow(vusb));
+                }
+                dataGridViewUsbList.Rows.AddRange(VmUsbRowToAdd.ToArray());
+
+                buttonDetach.Enabled = false;
+            }
+            finally
+            {
+                if (dataGridViewUsbList.SortedColumn != null)
+                {
+                    dataGridViewUsbList.Sort(
+                        dataGridViewUsbList.SortedColumn,
+                        dataGridViewUsbList.SortOrder == SortOrder.Ascending
+                            ? ListSortDirection.Ascending : ListSortDirection.Descending);
+                }
+                dataGridViewUsbList.ResumeLayout();
+                InBuildList = false;
+            }
+        }
+
+        public void ShowLocalValidationMessages()
+        {
+
+        }
+
+        public void Cleanup()
+        {

```

Comment:
Here is the place where we unregister any event listeners, so we should call DeregisterEvents() for every row.

Reply:
OK, will fix.